Improve performance of nanmean/nansum on large arrays

JamesWrigley opened the pull request:
This implements some of the things described in #54, namely implementations for nanmean/nansum using patterns from Base. The changes to the loop ordering in 1e8c7c8 could be applied to the other functions too but I was lazy :sloth: I think it'd be good if we could reduce the amount of code duplication in the future.

The changes unfortunately ended up being a lot more invasive than I hoped because of the extra `size_threshold` argument (naming is up for debate). I added some benchmarks as well (smaller is better for all columns):

<details>
<summary>AMD Epyc server</summary>

|                            | dirty            | main             | dirty / main  |
|:---------------------------|:----------------:|:----------------:|:-------------:|
| nanmean/big array, dim 1   | 2.92 ± 0.044 ms  | 2.98 ± 0.049 ms  | 0.979 ± 0.022 |
| nanmean/big array, dim 2   | 11.1 ± 0.037 ms  | 13.1 ± 0.8 ms    | 0.844 ± 0.051 |
| nanmean/big array, dim 3   | 14.4 ± 3.5 ms    | 18.2 ± 0.91 ms   | 0.792 ± 0.2   |
| nanmean/small array, dim 1 | 0.59 ± 0.011 μs  | 0.6 ± 0.01 μs    | 0.983 ± 0.025 |
| nanmean/small array, dim 2 | 0.97 ± 0.01 μs   | 0.98 ± 0.01 μs   | 0.99 ± 0.014  |
| nanmean/small array, dim 3 | 0.77 ± 0.01 μs   | 0.87 ± 0.002 μs  | 0.885 ± 0.012 |
| nanmean/small vector       | 31 ± 10 ns       | 30 ± 1 ns        | 1.03 ± 0.34   |
| nansum/big array, dim 1    | 2.75 ± 0.062 ms  | 2.83 ± 0.05 ms   | 0.972 ± 0.028 |
| nansum/big array, dim 2    | 2.82 ± 0.045 ms  | 10 ± 0.44 ms     | 0.282 ± 0.013 |
| nansum/big array, dim 3    | 3.26 ± 0.51 ms   | 18.6 ± 0.7 ms    | 0.175 ± 0.028 |
| nansum/small array, dim 1  | 0.589 ± 0.01 μs  | 0.54 ± 0.011 μs  | 1.09 ± 0.029  |
| nansum/small array, dim 2  | 0.551 ± 0.01 μs  | 0.611 ± 0.02 μs  | 0.902 ± 0.034 |
| nansum/small array, dim 3  | 0.53 ± 0.05 μs   | 0.72 ± 0.019 μs  | 0.736 ± 0.072 |
| nansum/small vector        | 30 ± 0 ns        | 30 ± 0 ns        | 1 ± 0         |
| time_to_load               | 0.189 ± 0.0022 s | 0.189 ± 0.0032 s | 1 ± 0.021     |

</details>

<details>
<summary>Intel server</summary>

|                            | dirty            | main               | dirty / main   |
|:---------------------------|:----------------:|:------------------:|:--------------:|
| nanmean/big array, dim 1   | 9.5 ± 0.6 ms     | 8.79 ± 0.51 ms     | 1.08 ± 0.093   |
| nanmean/big array, dim 2   | 14.2 ± 0.26 ms   | 0.0332 ± 8.9e-05 s | 0.429 ± 0.0078 |
| nanmean/big array, dim 3   | 21.4 ± 4.3 ms    | 0.0362 ± 0.0019 s  | 0.591 ± 0.12   |
| nanmean/small array, dim 1 | 1.27 ± 0.053 μs  | 1.3 ± 0.019 μs     | 0.975 ± 0.043  |
| nanmean/small array, dim 2 | 1.62 ± 0.017 μs  | 1.68 ± 0.017 μs    | 0.964 ± 0.014  |
| nanmean/small array, dim 3 | 1.5 ± 0.033 μs   | 1.44 ± 0.015 μs    | 1.04 ± 0.025   |
| nanmean/small vector       | 0.049 ± 0.003 μs | 0.051 ± 0.004 μs   | 0.961 ± 0.096  |
| nansum/big array, dim 1    | 8.4 ± 0.54 ms    | 8.02 ± 0.6 ms      | 1.05 ± 0.1     |
| nansum/big array, dim 2    | 8.35 ± 0.23 ms   | 25.6 ± 0.086 ms    | 0.326 ± 0.009  |
| nansum/big array, dim 3    | 13.8 ± 1.1 ms    | 24 ± 1.8 ms        | 0.573 ± 0.063  |
| nansum/small array, dim 1  | 0.911 ± 0.031 μs | 1.05 ± 0.006 μs    | 0.871 ± 0.03   |
| nansum/small array, dim 2  | 1.01 ± 0.071 μs  | 1.22 ± 0.019 μs    | 0.825 ± 0.06   |
| nansum/small array, dim 3  | 1.05 ± 0.12 μs   | 0.965 ± 0.007 μs   | 1.09 ± 0.13    |
| nansum/small vector        | 0.046 ± 0.001 μs | 0.044 ± 0.004 μs   | 1.05 ± 0.098   |
| time_to_load               | 0.258 ± 0.015 s  | 0.264 ± 0.0094 s   | 0.976 ± 0.068  |

</details>

<details>
<summary>Intel laptop</summary>

Nansum improvements are weirdly massive here for some reason.

|                            | dirty            | main               | dirty / main   |
|:---------------------------|:----------------:|:------------------:|:--------------:|
| nanmean/big array, dim 1   | 4.74 ± 0.051 ms  | 4.83 ± 0.043 ms    | 0.981 ± 0.014  |
| nanmean/big array, dim 2   | 10.2 ± 0.045 ms  | 0.0493 ± 0.00095 s | 0.208 ± 0.0041 |
| nanmean/big array, dim 3   | 14.3 ± 3.1 ms    | 0.0564 ± 0.002 s   | 0.254 ± 0.056  |
| nanmean/small array, dim 1 | 0.736 ± 0.081 μs | 0.768 ± 0.057 μs   | 0.958 ± 0.13   |
| nanmean/small array, dim 2 | 1.48 ± 0.14 μs   | 1.6 ± 0.21 μs      | 0.922 ± 0.15   |
| nanmean/small array, dim 3 | 1.64 ± 0.16 μs   | 1.69 ± 0.17 μs     | 0.965 ± 0.13   |
| nanmean/small vector       | 0.04 ± 0.003 μs  | 0.046 ± 0.002 μs   | 0.87 ± 0.075   |
| nansum/big array, dim 1    | 4.32 ± 0.07 ms   | 4.4 ± 0.061 ms     | 0.981 ± 0.021  |
| nansum/big array, dim 2    | 4.59 ± 0.07 ms   | 0.045 ± 0.00011 s  | 0.102 ± 0.0016 |
| nansum/big array, dim 3    | 9.23 ± 0.7 ms    | 0.0572 ± 0.0014 s  | 0.161 ± 0.013  |
| nansum/small array, dim 1  | 0.858 ± 0.074 μs | 0.959 ± 0.11 μs    | 0.895 ± 0.13   |
| nansum/small array, dim 2  | 0.937 ± 0.075 μs | 1.15 ± 0.11 μs     | 0.813 ± 0.1    |
| nansum/small array, dim 3  | 0.867 ± 0.073 μs | 1.26 ± 0.12 μs     | 0.687 ± 0.086  |
| nansum/small vector        | 0.04 ± 0.001 μs  | 0.043 ± 0.001 μs   | 0.93 ± 0.032   |
| time_to_load               | 0.154 ± 0.0079 s | 0.16 ± 0.0086 s    | 0.962 ± 0.071  |

</details>

I've tried to keep all the commits atomic so I'd recommend reviewing them one-by-one.

EDIT: two issues in CI:
- Something is very wrong on 32bit, maybe an overflow somewhere.
- ~~On nightly there's a nan in the output :grimacing:~~ Dumb mistake, fixed in 4fca8ce.

brenhinkeller commented on src/ArrayStats/nanmean.jl:
```diff
@@ -1,6 +1,11 @@
+# Default to 1MB
+NANMEAN_SIZE_THRESHOLD::Union{Int, Symbol} = 2^20
```
Are there any performance or type inference implications that we should keep an eye out for from having this as a type-annotated global vs a const integer global?

JamesWrigley replied:
I think we're safe, from looking at the function in Cthulhu it looks like the world-splitting optimization is kicking in:
```julia
   %31 = #nanmean#13(::Int64,::Colon,::Int64,::typeof(nanmean),::Matrix{Float64})::Matrix{Float64}
 • %31 = #nanmean#13(::Int64,::Colon,::Symbol,::typeof(nanmean),::Matrix{Float64})::Matrix{Float64}
```

And if I descend into the `Symbol` case the `get_size_threshold(::Symbol)` call gets inferred to an Int:
```julia
_nanmean(A::AbstractArray{T, N}, dims::Tuple, st) where {T, N} @ NaNStatistics ~/git/NaNStatistics/src/ArrayStats/nanmean.jl:55
55 function (_nanmean(A::Matrix{Float64}::AbstractArray{T,N}, dims::Tuple{Int64}::Tuple, st::Symbol) where {T,N})::Matrix{Float64}
56     sᵢ::Tuple{Int64, Int64} = size(A::Matrix{Float64})::Tuple{Int64, Int64}
57     sₒ::Tuple{Int64, Int64} = ntuple(Val{N::Core.Const(2)}::Type{Val{2}}()::Core.Const(Val{2}()))::Tuple{Int64, Int64} do d
58         ifelse(d ∈ dims, 1, sᵢ[d])
59     end
60     Tₒ::Type{Float64} = Base.promote_op(/::Core.Const(/), T::Type{Float64}, Int::Type{Int64})::Type{Float64}
61     B::Matrix{Float64} = similar(A::Matrix{Float64}, Tₒ::Type{Float64}, sₒ::Tuple{Int64, Int64})::Matrix{Float64}
63 
63     if (1 in dims::Tuple{Int64})::Bool || (sizeof(A::Matrix{Float64})::Int64 < get_size_threshold(st::Symbol)::Int64)::Bool
64         # The generated-function approach is faster for small arrays and if
65         # we're reducing over the first dimension.
66         _nanmean!(B::Matrix{Float64}, A::Matrix{Float64}, dims::Tuple{Int64}, st::Symbol)::Matrix{Float64}
67     else
68         # For reducing over the slow axes of large arrays we use the mapreduce
69         # approach.
70         _nanmean_mapreduce!(B::Matrix{Float64}, A::Matrix{Float64}, dims::Tuple{Int64})::Matrix{Float64}
71     end
72 end
```

JamesWrigley commented:
I've made a few changes in the latest version:
- Fixed the 32bit issue by initializing the output array first.
- Significantly simplified the implementation by removing the `size_threshold` argument from most of the functions, since they already allow taking a `mean` argument I figure people who want something other than the defaults can compute it themselves.

Updated benchmarks (they look much the same as before):

<details>
<summary>AMD Epyc server</summary>

|                            | dirty            | main             | dirty / main   |
|:---------------------------|:----------------:|:----------------:|:--------------:|
| nanmean/big array, dim 1   | 2.95 ± 0.048 ms  | 2.97 ± 0.035 ms  | 0.993 ± 0.02   |
| nanmean/big array, dim 2   | 10.8 ± 0.023 ms  | 13.4 ± 0.028 ms  | 0.806 ± 0.0024 |
| nanmean/big array, dim 3   | 13.5 ± 0.87 ms   | 18 ± 0.74 ms     | 0.749 ± 0.058  |
| nanmean/small array, dim 1 | 0.57 ± 0.01 μs   | 0.58 ± 0.019 μs  | 0.983 ± 0.037  |
| nanmean/small array, dim 2 | 0.99 ± 0.01 μs   | 0.98 ± 0.01 μs   | 1.01 ± 0.015   |
| nanmean/small array, dim 3 | 0.72 ± 0.01 μs   | 0.89 ± 0.011 μs  | 0.809 ± 0.015  |
| nanmean/small vector       | 30 ± 10 ns       | 30 ± 10 ns       | 1 ± 0.47       |
| nansum/big array, dim 1    | 2.79 ± 0.032 ms  | 2.82 ± 0.042 ms  | 0.988 ± 0.019  |
| nansum/big array, dim 2    | 2.85 ± 0.025 ms  | 10.6 ± 0.6 ms    | 0.27 ± 0.016   |
| nansum/big array, dim 3    | 3.49 ± 0.34 ms   | 17 ± 0.66 ms     | 0.206 ± 0.022  |
| nansum/small array, dim 1  | 0.53 ± 0.01 μs   | 0.54 ± 0.011 μs  | 0.981 ± 0.027  |
| nansum/small array, dim 2  | 0.55 ± 0.03 μs   | 0.61 ± 0.02 μs   | 0.902 ± 0.057  |
| nansum/small array, dim 3  | 0.531 ± 0.061 μs | 0.71 ± 0.01 μs   | 0.748 ± 0.087  |
| nansum/small vector        | 30 ± 0 ns        | 30 ± 0 ns        | 1 ± 0          |
| time_to_load               | 0.194 ± 0.0034 s | 0.198 ± 0.0065 s | 0.981 ± 0.037  |


</details>

<details>
<summary>Intel server</summary>

|                            | dirty            | main              | dirty / main   |
|:---------------------------|:----------------:|:-----------------:|:--------------:|
| nanmean/big array, dim 1   | 9.43 ± 0.59 ms   | 8.82 ± 0.32 ms    | 1.07 ± 0.077   |
| nanmean/big array, dim 2   | 13.7 ± 0.18 ms   | 0.0326 ± 0.0003 s | 0.418 ± 0.0068 |
| nanmean/big array, dim 3   | 17.3 ± 1.6 ms    | 0.0343 ± 0.0014 s | 0.504 ± 0.05   |
| nanmean/small array, dim 1 | 1.16 ± 0.022 μs  | 1.19 ± 0.016 μs   | 0.973 ± 0.023  |
| nanmean/small array, dim 2 | 1.5 ± 0.038 μs   | 1.53 ± 0.02 μs    | 0.98 ± 0.028   |
| nanmean/small array, dim 3 | 1.34 ± 0.01 μs   | 1.31 ± 0.014 μs   | 1.03 ± 0.013   |
| nanmean/small vector       | 0.047 ± 0.001 μs | 0.046 ± 0.001 μs  | 1.02 ± 0.031   |
| nansum/big array, dim 1    | 8.19 ± 0.77 ms   | 7.95 ± 0.61 ms    | 1.03 ± 0.12    |
| nansum/big array, dim 2    | 8.34 ± 0.18 ms   | 24.5 ± 0.051 ms   | 0.341 ± 0.0076 |
| nansum/big array, dim 3    | 13.7 ± 0.82 ms   | 22.7 ± 1.7 ms     | 0.603 ± 0.058  |
| nansum/small array, dim 1  | 0.827 ± 0.026 μs | 0.796 ± 0.008 μs  | 1.04 ± 0.034   |
| nansum/small array, dim 2  | 0.938 ± 0.083 μs | 1.27 ± 0.015 μs   | 0.742 ± 0.066  |
| nansum/small array, dim 3  | 0.857 ± 0.12 μs  | 1.05 ± 0.055 μs   | 0.815 ± 0.12   |
| nansum/small vector        | 0.042 ± 0.001 μs | 0.042 ± 0.001 μs  | 1 ± 0.034      |
| time_to_load               | 0.228 ± 0.001 s  | 0.217 ± 0.019 s   | 1.05 ± 0.093   |

</details>

<details>
<summary>Intel laptop</summary>

|                            | dirty            | main               | dirty / main     |
|:---------------------------|:----------------:|:------------------:|:----------------:|
| nanmean/big array, dim 1   | 4.69 ± 0.043 ms  | 4.75 ± 0.041 ms    | 0.987 ± 0.012    |
| nanmean/big array, dim 2   | 10.2 ± 0.045 ms  | 0.047 ± 0.00088 s  | 0.218 ± 0.0042   |
| nanmean/big array, dim 3   | 13.5 ± 2.5 ms    | 0.0582 ± 0.0015 s  | 0.231 ± 0.043    |
| nanmean/small array, dim 1 | 0.733 ± 0.12 μs  | 0.731 ± 0.059 μs   | 1 ± 0.18         |
| nanmean/small array, dim 2 | 1.49 ± 0.5 μs    | 1.51 ± 0.16 μs     | 0.985 ± 0.35     |
| nanmean/small array, dim 3 | 1.62 ± 0.2 μs    | 1.66 ± 0.16 μs     | 0.981 ± 0.15     |
| nanmean/small vector       | 0.044 ± 0.002 μs | 0.046 ± 0.001 μs   | 0.957 ± 0.048    |
| nansum/big array, dim 1    | 4.29 ± 0.088 ms  | 4.33 ± 0.044 ms    | 0.991 ± 0.023    |
| nansum/big array, dim 2    | 4.49 ± 0.041 ms  | 0.0451 ± 8.2e-05 s | 0.0995 ± 0.00093 |
| nansum/big array, dim 3    | 9.08 ± 0.83 ms   | 0.0572 ± 0.0014 s  | 0.159 ± 0.015    |
| nansum/small array, dim 1  | 0.844 ± 0.081 μs | 0.844 ± 0.086 μs   | 1 ± 0.14         |
| nansum/small array, dim 2  | 0.918 ± 0.1 μs   | 1.14 ± 0.11 μs     | 0.807 ± 0.12     |
| nansum/small array, dim 3  | 0.858 ± 0.12 μs  | 1.24 ± 0.11 μs     | 0.69 ± 0.11      |
| nansum/small vector        | 0.043 ± 0.002 μs | 0.043 ± 0.002 μs   | 1 ± 0.066        |
| time_to_load               | 0.154 ± 0.0011 s | 0.152 ± 0.00098 s  | 1.01 ± 0.0099    |

</details>

brenhinkeller commented:
Looks like a clear win on performance!
